test/json_unittest.cc

```diff
@@ -0,0 +1,385 @@
+}
+
+TEST(TrivialParseTest, ArrayOneElement) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("[2]");
+    const auto& arr = v->As<json::Array>();
+    EXPECT_EQ(arr->size(), 1);
+    EXPECT_EQ((*arr)[0]->As<json::Number>()->value(), 2.0);
+  });
+}
+
+TEST(TrivialToStringTest, ArrayOneElement) {
+  GuardJsonException([](){
+    EXPECT_TO_STRING(json::array({json::number(2.0)}), "[2.0]");
+  });
+}
+
+
+// Trival Dict
+
+TEST(TrivialParseTest, DictEmpty) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("{}");
+    EXPECT_TRUE(v->As<json::Object>()->empty());
+  });
+}
+
+TEST(TrivialToStringTest, DictEmpty) {
+  GuardJsonException([](){
+    EXPECT_TO_STRING(json::object({}), "{}");
+  });
+}
+
+TEST(TrivialParseTest, DictOneField) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("{\"f\":2}");
+    const auto& obj = v->As<json::Object>();
+    EXPECT_EQ(obj->size(), 1);
+    EXPECT_EQ(obj->Get<json::Number>("f"), 2.0);
+  });
+}
+
+TEST(TrivialToStringTest, DictOneField) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::object({
+      {"f", json::number(2.0)}
+    });
+    EXPECT_TO_STRING(v, "{\"f\":2.0}");
+  });
+}
+
+
+// String edge cases
+
+/*
+ * TODO(igorp): Seems that yajl does not support valid JSON.
+TEST(EdgeTest, StringWithNewLine) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"\n\"");
+    EXPECT_EQ(v->As<json::String>()->value(), "\n");
+    // Output should be escaped with \n as that is more canonical.
+    EXPECT_TO_STRING(v, "\"\n\"");
+  });
+}
+*/
+
+TEST(EdgeTest, StringWithNonUnicodeEscape) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"foo\\nbar\"");
+    EXPECT_EQ(v->As<json::String>()->value(), "foo\nbar");
+    // Output should be escaped as that is more canonical.
+    EXPECT_TO_STRING(v, "\"foo\\nbar\"");
+  });
+}
+
+TEST(EdgeTest, StringWithEveryEscape) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"x\\\\x\\/x\\bx\\fx\\nx\\rx\\tx\"");
+    EXPECT_EQ(v->As<json::String>()->value(), "x\\x/x\bx\fx\nx\rx\tx");
+    // Output should be escaped as that is more canonical.
+    // Since / does not need to be escaped, output does not escape that.
+    EXPECT_TO_STRING(v, "\"x\\\\x/x\\bx\\fx\\nx\\rx\\tx\"");
+  });
+}
+
+TEST(EdgeTest, StringWithUnicodeEscape) {
+  GuardJsonException([](){
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"foo\\u000abar\"");
+    //EXPECT_EQ(v->As<json::String>()->value(), "foo\nbar");
+    // Output should be escaped as that is more canonical.
+    EXPECT_TO_STRING(v, "\"foo\\nbar\"");
+  });
+}
+
+TEST(EdgeTest, StringWithUTF8) {
+  GuardJsonException([](){
+    // This is Korean for "Hello World!".
+    std::unique_ptr<json::Value> v = json::Parser::FromString("\"안녕 세상아!\"");
```

Comment:
My heartfelt thanks for not using the [poop emoji](http://www.fileformat.info/info/unicode/char/1F4A9/index.htm). 😄